routes/users-CODEREVIEWCOPY.js

```diff
@@ -0,0 +1,142 @@
+const express = require("express");
+const router = express.Router();
+const db = require("../db/models");
+const { asyncHandler, csrfProtection, handleValidationErrors } = require("../utils.js");
+const { check, validationResult, body } = require("express-validator");
+const bcrypt = require("bcryptjs");
+const { loginUser, logoutUser } = require("../auth");
+
+const userValidators = [
+  check("userName")
+    .exists({ checkFalsy: true })
+    .withMessage("Please provide a username")
+    .isLength({ max: 50 })
+    .withMessage("userName must not be longer than 50 characters"),
+  check("email")
+    .exists({ checkFalsy: true })
+    .withMessage("Please provide an email")
+    .isLength({ max: 255 })
+    .withMessage("Email must not be longer than 255 characters")
+    .isEmail()
+    .withMessage("Please provide a valid email address")
+    .custom((value) => {
+      return db.User.findOne({ where: { email: value } }).then((user) => {
+        if (user) {
+          return Promise.reject(
+            "The provided Email Address is already in use by another account"
+          );
+        }
+      });
+    }),
+  check("password")
+    .exists({ checkFalsy: true })
+    .withMessage("Please provide a password")
+    .isLength({ max: 50 })
+    .withMessage("Password must not be longer than 50 characters")
+    .matches(/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*])/, "g")
+    .withMessage(
+      'Password must contain at least 1 lowercase letter, uppercase letter, number, and special character (i.e. "!@#$%^&*")'
+    ),
+  check("confirmPassword")
+    .exists({ checkFalsy: true })
+    .withMessage("Please confirm your password")
+    .isLength({ max: 50 })
+    .withMessage("Password must not be longer than 50 characters")
+    .custom((value, { req }) => {
+      if (value !== req.body.password) {
+        throw new Error("Confirm Password does not match Password");
+      }
+      return true;
+    }),
+];
+
+const loginValidators = [
+check("userName")
+  .exists({ checkFalsy: true })
+  .withMessage("Please enter your user name."),
+check("password")
+  .exists({ checkFalsy: true })
+  .withMessage("Please enter your password."),
+];
+
+/* GET signup page. */
+router.get(
+  "/signup",
+  csrfProtection,
+  asyncHandler(async (req, res, next) => {
+    const user = db.User.build();
+    res.render("signup", {
+      title: "Sign up",
+      user,
+      csrfToken: req.csrfToken(),
+    });
+  })
+);
+
+// Sign Up Users POST
+router.post(
+  "/signup",
+  csrfProtection,
+  userValidators,
+  asyncHandler(async (req, res) => {
+  const {userName, email, password} = req.body
+  const user = db.User.build({userName, email})
+  const validatorErrors = validationResult(req)
+  if(validatorErrors.isEmpty()){
+    const hashedPassword = await bcrypt.hash(password, 10)
+    user.hashedPassword = hashedPassword
+    console.log(req.body)
```

Comment:
Don't forget to remove your `console.log`s before you merge work into your master/main branch. You don't want noisy logs building up and hurting your debugging.